src/OrleansAWSUtils/Providers/Membership/OrleansSiloInstanceManager.cs

```diff
@@ -0,0 +1,278 @@
+
+            if (numTableVersionRows > 1)
+                throw new KeyNotFoundException(string.Format("Read {0} table version rows, while was expecting only 1. Read = {1}", numTableVersionRows, Utils.EnumerableToString(asList)));
+
+            return asList;
+        }
+
+        internal async Task<List<SiloInstanceTableEntry>> FindAllSiloEntries()
+        {
+            var asList = await storage.ReadAllTableEntriesForPartitionAsync(this.DeploymentId);
+            if (asList.Count < 1)
+                throw new KeyNotFoundException(string.Format("Could not find enough rows in the FindAllSiloEntries call. Found = {0}", Utils.EnumerableToString(asList)));
+
+            int numTableVersionRows = asList.Count(r => r.RowKey == SiloInstanceTableEntry.TABLE_VERSION_ROW);
+            if (numTableVersionRows < 1)
+                throw new KeyNotFoundException(string.Format("Did not find table version row. Read = {0}", Utils.EnumerableToString(asList)));
+            if (numTableVersionRows > 1)
+                throw new KeyNotFoundException(string.Format("Read {0} table version rows, while was expecting only 1. Read = {1}", numTableVersionRows, Utils.EnumerableToString(asList)));
+
+            return asList;
+        }
+
+        /// <summary>
+        /// Insert (create new) row entry
+        /// </summary>
+        internal async Task<bool> TryCreateTableVersionEntryAsync()
+        {
+            var versionRow = await storage.ReadSingleTableEntryAsync(DeploymentId, SiloInstanceTableEntry.TABLE_VERSION_ROW);
+            if (versionRow != null)
+            {
+                return false;
+            }
+            SiloInstanceTableEntry entry = CreateTableVersionEntry(0);
+            await storage.UpsertItemAsync(entry);
+            return true;
+        }
+
+        /// <summary>
+        /// Insert (create new) row entry
+        /// </summary>
+        /// <param name="siloEntry">Silo Entry to be written</param>
+        /// <param name="tableVersionEntry">Version row to update</param>
+        internal async Task<bool> InsertSiloEntryConditionally(SiloInstanceTableEntry siloEntry, SiloInstanceTableEntry tableVersionEntry)
+        {
+            await storage.UpsertItemsAsync(new[] { siloEntry, tableVersionEntry });
+            return true;
+        }
+
+        /// <summary>
+        /// Conditionally update the row for this entry, but only if the version matches with the current record in data store
+        /// </summary>
+        /// <param name="siloEntry">Silo Entry to be written</param>
+        /// <param name="tableVersionEntry">Version row to update</param>
+        /// <returns></returns>
+        internal async Task<bool> UpdateSiloEntryConditionally(SiloInstanceTableEntry siloEntry, SiloInstanceTableEntry tableVersionEntry)
+        {
+            await storage.UpsertItemAsync(tableVersionEntry);
+            await storage.UpsertItemAsync(siloEntry);
```

Comment:
This should not be upsert. It should be update conditional conditioned on stage/version. Also the two updates should be applied atomically, in the same batch transaction.